sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_outlier_filter/cuda_polar_voxel_outlier_filter.cu

```diff
@@ -0,0 +1,949 @@
+  //     is_valid = is_primary_returns[point_index.value()];
+  //   }
+  // }
+  // valid_points_mask[point_index.value()] = is_valid;
+  // -----
+
+  valid_points_mask[point_index.value()] =
+    meet_voxel_criteria && (!filter_secondary_returns || is_primary_returns[point_index.value()]);
+}
+
+__global__ void copy_valid_points_kernel(
+  const uint8_t * __restrict__ input_cloud, const bool * __restrict__ valid_points_mask,
+  const int * __restrict__ filtered_indices, const size_t num_points, const size_t step,
+  uint8_t * __restrict__ output_points)
+{
+  auto point_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (point_index >= num_points) {
+    return;
+  }
+
+  bool is_valid = valid_points_mask[point_index];
+  int destination_index = filtered_indices[point_index];
+
+  if (is_valid) {
+    memcpy(output_points + destination_index * step, input_cloud + point_index * step, step);
+  }
+}
+
+__global__ void bool_flip_kernel(bool * __restrict__ flags, const size_t num_points)
+{
+  auto point_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (point_index >= num_points) {
+    return;
+  }
+
+  flags[point_index] = !flags[point_index];
+}
+
+__global__ void is_low_visibility_voxel_kernel(
+  const bool * __restrict__ secondary_meets_threshold,
+  const int * __restrict__ is_in_visibility_range, const int num_voxels,
+  bool * __restrict__ is_low_visibility_voxels)
+{
+  auto voxel_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (voxel_index >= num_voxels) {
+    return;
+  }
+
+  is_low_visibility_voxels[voxel_index] = static_cast<bool>(is_in_visibility_range[voxel_index]) &&
+                                          !secondary_meets_threshold[voxel_index];
+}
+
+// Helper function to get field offset
+template <typename T>
+size_t get_offset(const T & fields, const std::string & field_name)
+{
+  int index = -1;
+  for (size_t i = 0; i < fields.size(); ++i) {
+    if (fields[i].name == field_name) {
+      index = static_cast<int>(i);
+      break;
+    }
+  }
+  if (index < 0) {
+    std::stringstream ss;
+    ss << "input cloud does not contain field named '" << field_name << "'";
+    throw std::runtime_error(ss.str());
+  }
+  return fields[index].offset;
+}
+}  // namespace
+
+CudaPolarVoxelOutlierFilter::CudaPolarVoxelOutlierFilter() : primary_return_type_dev_(std::nullopt)
+{
+  CHECK_CUDA_ERROR(cudaStreamCreate(&stream_));
+
+  // create memory pool to make repeated allocation efficient
+  int current_device_id = 0;
+  CHECK_CUDA_ERROR(cudaGetDevice(&current_device_id));
+  size_t max_mem_pool_size_in_byte = 1e9;  // 1GB
+  mem_pool_ =
+    autoware::cuda_utils::create_memory_pool(max_mem_pool_size_in_byte, current_device_id);
+}
+
+CudaPolarVoxelOutlierFilter::FilterReturn CudaPolarVoxelOutlierFilter::filter(
+  const cuda_blackboard::CudaPointCloud2::ConstSharedPtr & input_cloud,
+  const CudaPolarVoxelOutlierFilterParameters & params, const PolarDataType polar_type)
+{
+  if (!primary_return_type_dev_) {
```

Comment:
This is guaranteed to be set in the node constructor, and when changing params at runtime. Thus, this should probably be handled as an error here.